### PR TITLE
Integrating OneDAL primitives for `LogisticRegression`

### PR DESCRIPTION
**Pull request by avolkov-intel**
# Description
#### Add scikit-learn-intelex interface for oneDAL algorithm objective_function
- Defined new class LogisticLoss (defined in onedal/objective_function/objective_function.py). Its interface is almost alligned to LinearModelLoss class from stock scikit-learn https://github.com/scikit-learn/scikit-learn/blob/7db5b6a98ac6ad0976a3364966e214926ca8098a/sklearn/linear_model/_linear_loss.py#L9
- Defined functions loss, gradient, loss_gradient, gradient_hessian, gradient_hessian_product
- Covered new functionality with tests

### CODE REVIEW

File: onedal/__init__.py [hunk 1]
```diff
@@ -46,7 +46,7 @@
     __all__.append('spmd')
 
 if daal_check_version((2023, 'P', 100)):
-    __all__ += ['basic_statistics', 'linear_model']
+    __all__ += ['basic_statistics', 'linear_model', 'objective_function']
```
Comment (samir-nasibli) [inline onedal/__init__.py:49]:
```suggestion
    __all__ += ['basic_statistics', 'linear_model', 'metrics']
```
I suppose it should be in `metrics` module

**Comment by samir-nasibli** [comment]
I really don't understand which api are your implementing here? in vanilla scikit-learn we already have API for log_loss https://scikit-learn.org/stable/modules/generated/sklearn.metrics.log_loss.html . I think it's worth following this API

File: onedal/__init__.py (hunk 1 above)
Comment (avolkov-intel) [reply onedal/__init__.py:49]: Metric is the name of function that we use to measure the quality of our predictions, while objective function is something that we use to fit the algorithm (calculate gradient, do optimization steps and etc.) Moreover, all the modules in onedal are already named objective_function so if we use metrics here this would lead to confusion. So I would not agree with this comment

**Comment by avolkov-intel** [comment]
> I really don't understand which api are your implementing here? in vanilla scikit-learn we already have API for log_loss https://scikit-learn.org/stable/modules/generated/sklearn.metrics.log_loss.html . I think it's worth following this API

I implement the API for LinearModelLoss class. https://github.com/scikit-learn/scikit-learn/blob/7db5b6a98ac6ad0976a3364966e214926ca8098a/sklearn/linear_model/_linear_loss.py#L9
This class is used internally by scikit-learn LogisticRegression

File: onedal/__init__.py (hunk 1 above)
Comment (samir-nasibli) [reply onedal/__init__.py:49]: The `metric` proposed here is just module name, not a function, like in scikit-learn, that includes `log_loss and other classification and regression metric functions.

We don't have any naming rules here. For example, in sklearnex/onedal we are mostly following scikit-learn like interface, so naming as well.

**Comment by avolkov-intel** [comment]
@KulikovNikita @Alexsandruss could you review the code, please?

File: onedal/objective_function/objective_function.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+    ONEDAL_PY_INSTANTIATE(init_compute_ops, sub, policy_list, task_list);
+    ONEDAL_PY_INSTANTIATE(init_compute_result, sub, task_list);
+}
+
+ONEDAL_PY_TYPE2STR(dal::objective_function::task::compute, "compute");
+
+#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20230100
+
+} // namespace oneapi::dal::python
```
Comment (samir-nasibli) [inline onedal/objective_function/objective_function.cpp:179]: add extra blank line

File: onedal/objective_function/objective_function.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+
+class LogisticLoss(BaseObjectiveFunction):
+    def __init__(
+        self, *, algorithm="by_default", queue=None, fit_intercept=True, **kwargs
+    ):
+        self.fit_intercept = fit_intercept
+        super().__init__(algorithm, queue, _backend.objective_function.compute.logloss)
+
+    def __fix_gradient(self, grad, coef, l2_reg_strength):
```
Comment (samir-nasibli) [inline onedal/objective_function/objective_function.py:115]: Why are you using double pre underscore in functions names in this class?

File: onedal/objective_function/objective_function.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        X,
+        y,
+        sample_weight=None,
+        l2_reg_strength=0.0,
+        n_threads=1,
+        raw_prediction=None,
+    ):
+        assert sample_weight is None
+        assert n_threads == 1
+        assert raw_prediction is None
```
Comment (samir-nasibli) [inline onedal/objective_function/objective_function.py:242]: Here should be error throw with related messages.

File: onedal/objective_function/objective_function.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        super().__init__(algorithm, queue, _backend.objective_function.compute.logloss)
+
+    def __fix_gradient(self, grad, coef, l2_reg_strength):
+        if self.fit_intercept:
+            grad = np.hstack([grad[1:] + coef[:-1] * l2_reg_strength, grad[0]])
+        else:
+            grad = grad[1:] + coef * l2_reg_strength
+        return grad
+
+    def __fix_hessian(self, hess, coef, l2_reg_strength):
```
Comment (samir-nasibli) [inline onedal/objective_function/objective_function.py:122]: I don't understand `fix` in the name of function. What is the purpose?

File: onedal/dal.cpp [hunk 6]
```diff
@@ -44,6 +44,7 @@ ONEDAL_PY_INIT_MODULE(decomposition);
 #if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20230100
 ONEDAL_PY_INIT_MODULE(basic_statistics);
 ONEDAL_PY_INIT_MODULE(linear_model);
+ONEDAL_PY_INIT_MODULE(objective_function);
```
Comment (samir-nasibli) [inline onedal/dal.cpp:47]: please add correct onedal versioning here

File: onedal/objective_function/tests/test_objective_function.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===============================================================================
+
+from daal4py.sklearn._utils import daal_check_version, sklearn_check_version
+from sklearn.metrics import log_loss
+from sklearn._loss.loss import HalfBinomialLoss
+from sklearn.linear_model import _linear_loss
+
+if daal_check_version((2023, "P", 100)):
```
Comment (samir-nasibli) [inline onedal/objective_function/tests/test_objective_function.py:22]: You already checking versioning and adding these test on build step. no need for these

File: onedal/objective_function/objective_function.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+
+    so to align with python interface format should be changed
+    '''
+
+    def __change_gradient_format(self, grad, coef, l2_reg_strength):
+        if self.fit_intercept:
+            grad = np.hstack([grad[1:] + coef[:-1] * l2_reg_strength, grad[0]])
+        else:
+            grad = grad[1:] + coef * l2_reg_strength
+        return grad
```
Comment (samir-nasibli) [inline onedal/objective_function/objective_function.py:131]: No need for double underscore in function name here.

File: onedal/objective_function/tests/test_objective_function.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+
+        product_onedal = hessp_onedal(s)
+        product_gth = hessp_gth(s)
+
+        assert_allclose(product_onedal, product_gth, rtol=tol)
+
+
+@pytest.mark.parametrize("queue", get_queues())
+@pytest.mark.parametrize(
+    "dimensions", [(10**6, 50), (100, 100), (500, 20), (20, 73)]
```
Comment (samir-nasibli) [inline onedal/objective_function/tests/test_objective_function.py:193]: First data volume is big for the test case.

File: onedal/objective_function/tests/test_objective_function.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+)
+def test_sizes(queue, dimensions):
+    tol = 1e-7
```
Comment (samir-nasibli) [inline onedal/objective_function/tests/test_objective_function.py:196]: no need in local variable here. Just pass value in `assert_allclose` param.

File: onedal/objective_function/objective_function.cpp [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+#include <string>
+#include <regex>
+#include <map>
+#include <iostream>
+
+namespace py = pybind11;
+
+namespace oneapi::dal::python {
+
+#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20230100
```
Comment (samir-nasibli) [inline onedal/objective_function/objective_function.cpp:31]:
```suggestion
#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20230200
```

File: onedal/objective_function/objective_function.cpp [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    Ops ops;
+};
+
+const std::map<std::string, dal::objective_function::result_option_id> result_option_registry{
+    { "value", dal::objective_function::detail::get_value_id() },
+    { "gradient", dal::objective_function::detail::get_gradient_id() },
+    { "hessian", dal::objective_function::detail::get_hessian_id() }
+};
```
Comment (samir-nasibli) [inline onedal/objective_function/objective_function.cpp:56]: Use function with switch cases instead of global variable here.

File: onedal/objective_function/objective_function.py [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+        if self.fit_intercept:
+            hess = np.hstack(
+                (
+                    np.vstack(
+                        [hess[1:, 1:] + np.diag([l2_reg_strength] * (num_params - 1)),
+                         hess[0, 1:],]
+                    ),
+                    np.hstack([hess[0, 1:], hess[0][0]]).reshape(-1, 1),
+                )
+            )
```
Comment (KulikovNikita) [inline onedal/objective_function/objective_function.py:165]: Can you please split it in steps?

File: onedal/objective_function/objective_function.py [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+        else:
+            hess = hess[1:, 1:] + np.diag([l2_reg_strength] * (num_params - 1))
+        return hess
+
+    def __calculate_regularization(self, coef, l2_reg_strength):
```
Comment (KulikovNikita) [inline onedal/objective_function/objective_function.py:170]: What about L1 regularization?

File: onedal/objective_function/objective_function.py [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+        if (n_threads != 1):
+            raise Exception("multithreading is not supported")
+        if (raw_prediction is not None):
+            raise Exception("raw_prediction parameter is not supported")
+
+        res = super()._compute(
+            X, y, coef, ["value", "gradient"], 0.0, self.fit_intercept, queue
+        )
+        value = res["value"]
+        grad = res["gradient"]
```
Comment (KulikovNikita) [inline onedal/objective_function/objective_function.py:223]:
```suggestion
        value, grad = res["value"], res["gradient"]
```

File: onedal/objective_function/tests/test_objective_function.py [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+
+        alg_internal = _linear_loss.LinearModelLoss(
+            base_loss=HalfBinomialLoss(), fit_intercept=fit_intercept,
+        )
+
+        logloss_gth = alg_internal.loss(
+            coef, data, y_true, l2_reg_strength=l2_reg_strength
+        )
+
+        print(logloss_onedal, logloss_gth)
```
Comment (KulikovNikita) [inline onedal/objective_function/tests/test_objective_function.py:70]: Please remove print statement before merging

File: onedal/objective_function/tests/test_objective_function.py [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+
+        alg_internal = _linear_loss.LinearModelLoss(
+            base_loss=HalfBinomialLoss(), fit_intercept=fit_intercept
+        )
+        gradient_gth = alg_internal.gradient(
+            coef, data, y_true, l2_reg_strength=l2_reg_strength
+        )
+
+        print(gradient_onedal)
+        print(gradient_gth)
```
Comment (KulikovNikita) [inline onedal/objective_function/tests/test_objective_function.py:96]: The same

File: onedal/objective_function/objective_function.py (hunk 14 above)
Comment (avolkov-intel) [reply onedal/objective_function/objective_function.py:179]: There is no l1-regularization in LinearModelLoss class and the interface should align with it.

File: onedal/objective_function/objective_function.py [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+        y,
+        sample_weight=None,
+        l2_reg_strength=0.0,
+        n_threads=1,
+        raw_prediction=None,
+        queue=None
+    ):
+        if (sample_weight is not None):
+            if (not np.array_equal(sample_weight, np.ones_like(sample_weight))):
+                raise Exception("sample_weigth parameter is not supported")
```
Comment (KulikovNikita) [inline onedal/objective_function/objective_function.py:218]: Please combine two checks

File: onedal/objective_function/tests/test_objective_function.py [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+            product_gth = hessp_gth(s)
+
+            assert_allclose(product_onedal, product_gth, rtol=tol)
+
+    @pytest.mark.parametrize("queue", get_queues())
+    @pytest.mark.parametrize(
+        "dimensions", [(100000, 15), (1000, 50), (100, 100), (500, 20), (20, 73)]
+    )
+    def test_sizes(queue, dimensions):
+        dtype = np.float64
```
Comment (KulikovNikita) [inline onedal/objective_function/tests/test_objective_function.py:189]: Please parametrize this test with types of variables

File: sklearnex/preview/__init__.py [hunk 20]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env python
-#===============================================================================
+# ===============================================================================
```
Comment (KulikovNikita) [inline sklearnex/preview/__init__.py:2]:
```suggestion
#===============================================================================
```

File: sklearnex/preview/__init__.py [hunk 21]
```diff
@@ -13,6 +13,6 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-#===============================================================================
+# ===============================================================================
```
Comment (KulikovNikita) [inline sklearnex/preview/__init__.py:16]:
```suggestion
#===============================================================================
```

File: sklearnex/preview/objective_function/tests/test_preview_logloss.py [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+    gradient3, hessian, flag = onedal_loss.gradient_hessian(coef2, X, y)
+
+    gradient4, hessp = onedal_loss.gradient_hessian_product(coef2, X, y)
+
+    logloss_gth = 10.129922567126792
+
+    gradient_gth = np.array(
+        [1.52457206, 1.81682052, -1.06863193, 0.81881473, -1.34379483])
+
+    hessian_gth = np.array([[5.18586773, 1.13656308, -1.32822742, 0.58710304, 0.37052724],
```
Comment (KulikovNikita) [inline sklearnex/preview/objective_function/tests/test_preview_logloss.py:55]: What about checking conformance with a vanilla sklearn?

**Review by KulikovNikita — COMMENTED** [review]
Generally, looks fine to me

**Review by KulikovNikita — APPROVED** [review]

File: onedal/objective_function/objective_function.py [hunk 23]
```diff
@@ -0,0 +1,314 @@
...
+        X,
+        y,
+        sample_weight=None,
+        l2_reg_strength=0.0,
+        n_threads=1,
+        raw_prediction=None,
+        queue=None
+    ):
+        if (sample_weight is not None):
+            if (not np.array_equal(sample_weight, np.ones_like(sample_weight))):
```
Comment (samir-nasibli) [inline onedal/objective_function/objective_function.py]: You are allocating extra array just for check? You should just check elements of sample_weight

**Comment by samir-nasibli** [comment]
Please rebase and run intel CI if it is required

**Review by samir-nasibli — CHANGES_REQUESTED** [review]
Please address comments

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

File: sklearnex/preview/objective_function/objective_function.py [hunk 24]
```diff
@@ -0,0 +1,287 @@
...
+                    return super().gradient_hessian_product(
+                        coef,
+                        X,
+                        y,
+                        sample_weight,
+                        l2_reg_strength,
+                        n_threads,
+                        raw_prediction)
+    else:
+        from sklearn.linear_model._linear_loss import LinearModelLoss
```
[the remark this answers is not in the document]
Comment (avolkov-intel) [reply sklearnex/preview/objective_function/objective_function.py:285]: Actually we can not avoid it because we want to patch only LinearModelLoss class but not the whole LogisticRegression algorithm

**Comment by ethanglaser** [comment]
Can reopen if progress continues
